contracts/meta-staking/src/contract.rs

```diff
@@ -1,55 +1,417 @@
+        // We subtract the amount delegated to the validator.
+        VALIDATORS_BY_CONSUMER.update(
+            deps.storage,
+            (&consumer_addr, &validator_addr),
+            |validator_info| -> Result<ValidatorInfo, ContractError> {
+                match validator_info {
+                    Some(validator_info) => Ok(ValidatorInfo {
+                        address: validator_info.address,
+                        consumer: validator_info.consumer,
+                        total_delegated: validator_info.total_delegated - amount.amount,
+                    }),
+                    // No one has delegated to this validator, throw error
+                    None => Err(ContractError::NoDelegationsForValidator {}),
+                }
+            },
+        )?;
+        // Increase the amount of available funds for that consumer
+        CONSUMERS.update(deps.storage, &info.sender, |current| match current {
+            Some(current) => Ok(ConsumerInfo {
+                address: current.address,
+                available_funds: available_funds + amount.amount,
+                total_staked: current.total_staked,
+            }),
+            None => Err(ContractError::Unauthorized {}),
+        })?;
+
+        // Create message to delegate the underlying tokens
+        let msg = StakingMsg::Undelegate { validator, amount };
+
+        Ok(Response::default().add_message(msg))
+    }
+
+    // TODO finish me
+    pub fn withdraw_delegator_reward(
+        deps: DepsMut,
+        _env: Env,
+        info: MessageInfo,
+        validator: String,
+    ) -> Result<Response, ContractError> {
+        // Check this is a consumer calling this, fails if no consumer loads
+        CONSUMERS.has(deps.storage, &info.sender);
+
+        // TODO Need to figure out how many rewards we got, so can send them
+        // to the consumer contract
+
+        // Withdraw rewards as a submessage
+        let withdraw_msg = SubMsg::reply_on_success(
+            DistributionMsg::WithdrawDelegatorReward { validator },
+            WITHDRAW_REWARDS_REPLY_ID,
+        );
+
+        // TODO On reply, send funds to consumer contract
+
+        Ok(Response::default().add_submessage(withdraw_msg))
+    }
+}
+
+// TODO query this info by consumer...
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::AllDelegations { delegator } => query::all_delegations(deps, delegator),
+        QueryMsg::AllValidators {} => query::all_validators(deps),
+        QueryMsg::BondedDenom {} => query::bonded_denom(deps),
+        QueryMsg::Config {} => query::config(deps),
+        QueryMsg::Delegation {
+            delegator,
+            validator,
+        } => query::delegation(deps, delegator, validator),
+        QueryMsg::Validator { address } => query::validator(deps, address),
+    }
+}
+
+mod query {
```

Comment:
Later: let's rethink these... maybe we make some more adjustments

Reply:
Yeah, these need work. They are just wrappers around standard stuff anyway.